src/atomics.h

```diff
@@ -43,6 +43,10 @@
  *        which is very hard unless the layout of the object is fully
  *        specified.
  */
+
+// TODO: volatile does not guarantee atomicity according to C++ standard
+typedef volatile unsigned long jl_atomic_ulong;
```

Comment:
There's no todo here. The lock does not rely on `volatile` for anything. In fact, the `volatile` can/should be removed.

Reply:
I understand it is working now on supported architectures.. but for better compliance with the standards and for sake of portability, it's better to have specifically designed type here - e.g. atomic_ulong from <stdatomic.h>. Unfortunately, according to the comment below, it needs more work or dropping gcc4.7 in order to implement. So, I wanted to express it here with this TODO, I can leave a comment without TODO though.

Reply:
Agree if these vars are only accessed through functions implementing atomic operations since an atomic type is only meaningful for lvalue operations. However, for a reader who don't dig that deep, it might not be obvious if the type is just volatile and has no any reference to its atomicity. Thus I wanted to introduce this special type in the first place, to hint that atomic access is needed here. Having proper atomic type will also enable atomic lvalue operations so we don't break expectations when somebody can rightfully assume that using lvalue operations for an atomic type is a legitimate thing.

Reply:
> Having proper atomic type

Please don't. That's why I have that long comment about cross-language support.

Reply:
no, worries, I don't want to mix this PR with huge refactoring of atomics.h and fixing all the compiler compatibility issues. I just want to introduce own Julia type for the sake described above.